core/capability-negotiation.md

```diff
@@ -0,0 +1,437 @@
+can delay capability negotiation until it has seen all available server capabilities.
+
+If a multi-line response to `CAP LIST` is sent, and the client supports CAP Version `302`,
+the server MUST delimit all replies except for the last one sent as noted above.
+
+Example with a client that **does not** support CAP version `302`:
+
+    Client: CAP LS
+    Server: CAP * LS :multi-prefix extended-join account-notify batch invite-notify tls
+    Server: CAP * LS :cap-notify server-time example.org/dummy-cap=dummyvalue example.org/second-dummy-cap
+    Server: CAP * LS :userhost-in-names sasl=EXTERNAL,DH-AES,DH-BLOWFISH,ECDSA-NIST256P-CHALLENGE,PLAIN
+
+Example with a client that supports CAP version `302`:
+
+    Client: CAP LS 302
+    Server: CAP * LS * :multi-prefix extended-join account-notify batch invite-notify tls
+    Server: CAP * LS * :cap-notify server-time example.org/dummy-cap=dummyvalue example.org/second-dummy-cap
+    Server: CAP * LS :userhost-in-names sasl=EXTERNAL,DH-AES,DH-BLOWFISH,ECDSA-NIST256P-CHALLENGE,PLAIN
+
+Example of a `LIST` reply with a client that supports CAP version `302`:
+
+    Client: CAP LIST
+    Server: CAP modernclient LIST * :example.org/example-cap example.org/second-example-cap account-notify
+    Server: CAP modernclient LIST :invite-notify batch example.org/third-example-cap
+
+### The CAP LIST subcommand
+
+The LIST subcommand is used to list the capabilities associated with the active connection.
+The client should send a LIST subcommand with no other arguments to solicit a list of
+active capabilities.
+
+If no capabilities are active, an empty parameter must be sent.
+
+Example:
+
+    Client: CAP LIST
+    Server: CAP * LIST :multi-prefix
+
+### The CAP REQ subcommand
+
+The REQ subcommand is used to request a change in capabilities associated with the active
+connection.  Its sole parameter must be a list of space-separated capability identifiers.
+Each capability identifier may be prefixed with a dash (`-`) to designate that the capability
+should be disabled.
+
+If a client requests a capability which is already enabled, or tries to disable a capability
+which is not enabled, the server MUST continue processing the REQ subcommand as though
+handling this capability was successful.
+
+The capability identifier set must be accepted as a whole, or rejected entirely.
+
+If a client issues a REQ subcommand, registration must be suspended until an END subcommand
+is received.
+
+Example:
+
+    Client: CAP REQ :multi-prefix sasl
+    Server: CAP * ACK :multi-prefix sasl
+
+### The CAP ACK subcommand
+
+The ACK subcommand is sent by the server to acknowledge a client-sent REQ, and let the client
+know that their requested capabilities have been enabled.
+
+This subcommand has a single parameter consisting of a space-separated list of capability
+names. Each capability name may be prefixed with a dash (`-`), indicating that this
+capability has been disabled as requested.
```

Comment:
This removes the following bit from the current spec:

> If an ACK reply originating from the server is spread across multiple lines, a client MUST NOT change capabilities until the last ACK of the set is received. Equally, a server MUST NOT change the capabilities of the client until the last ACK of the set has been sent.

Was that deliberate? As described it doesn't really make sense to me or seem possible to implement, so fair enough to drop it, but it needs an errata to explain why. If you removed it by mistake, any reason we should keep it?

Reply:
Yep, removed for the reasons you've stated. I'll add an errata to explain why

Reply:
Errata added to explain why that advice has been removed.

Reply:
I somewhat disagree with this, because of the following in 3.1 spec:

> The capability identifier set must be accepted as a whole, or rejected entirely.

Client knows which capabilities it requested in REQ, so it can see that some of them are missing in ACK. If some caps are missing, it should wait for the next ACK.

Reply:
I don't really expect clients to keep that meticulous of an eye on what caps they send and get back, and then compare the lists together before enabling them. I think that expecting clients to do so, while a nice idea, won't actually end up being done in reality, and so having that paragraph isn't useful. Thoughts?

Reply:
Hmm, well I guess the client is supposed to send END after ACK, so really it should probably wait for all REQed caps to be acked.

Reply:
Regarding usefulness of this feature: I don't really see how it's useful, so this is a mere speculation:  maybe server needs to do something asynchronously to enable a certain cap? In that case server would need to wait until all caps are enabled before it can send ACK. Now, it can ACK caps when they are prepared to be enabled. And after that, enable them when the last cap is asynchronously prepared to be enabled... So server needs to track the state anyway, uhh.

Well, this sentence is one of the reasons why ZNC requests capabilities from server one by one, waiting for ACK every time. Another reason is because it wants to know which capabilities are not enableable.

I'm mostly disagree here with bundling of two different things together: merge specs together to make them more understandable, and changing actual things (errata). If we want some errata, let's unbundle that in a separate pull request.

Reply:
So ZNC doesn’t trust that the caps listed in LS can be REQed? Seems bad to need to do it one by one. Or do you REQ blindly without LS?

Some clarifications will involve errata, but I don’t think these involve material changes to to the spec that should be relied on by implementations.

Reply:
No, it starts with LS. But it's possible that the capability was removed from server after it sent response to LS, in which case REQuesting all LSed caps will result in NAK. And at that point cap-notify is not enabled yet, because ZNC doesn't request CAP LS 302, as of now.

This change won't break ZNC because it requests them one by one, but it may break other implementations which do rely on this behavior.

Reply:
Here's a contrived example of such breakage:
cap `example/double` causes all characters to be doubled, e.g. `PPRRIIVVMMSSGG`
```
> CAP REQ :example/double multi-prefix
< CAP * ACK :example/double
< CCAAPP  **  AACCKK  ::mmuullttii--pprreeffiixx
```
Because client was waiting for all caps to be ACKed before using them, it won't understand second CAP ACK, which server sent using the capability which it enabled already.

Reply:
I think my main issue with the original statement isn't the usefulness of it – it's clearly useful for the server to be able to send all the `ACK` messages, and then after those are sent enable all the capabilities contained within. I think my main issue is just how difficult it is for clients to try to discover which `ACK`s it should respond to now, and which ones it should wait to enable when another message comes through.

As a client, I enable caps when I receive an `ACK` for those caps. Storing a list of the REQ's I've made, comparing them to the incoming `ACK`s and `NAK`s, and then waiting to enable stuff based on that seems somewhat fragile and is something that as a client I'd just avoid altogether by doing something like the above instead.

Reply:
Yep, I agree, and I chose to enable them one by one instead, which is not as fragile.

I'm just saying that if you're changing this part of spec, new servers won't be compatible with some old clients anymore, if those clients did implement this feature.

Reply:
Regarding the bundling, originally I didn't consider this a change, more just a clarification of what's actually done in the real world. But if there's contention I can break it out into a separate PR.

Reply:
Coolcool, I'll re-add this in and we'll discuss it more in a separate PR